cmd/execution_builder.go

```diff
@@ -644,6 +646,55 @@ func (exeNode *ExecutionNode) LoadExecutionState(
 	return &module.NoopReadyDoneAware{}, nil
 }
 
+func (exeNode *ExecutionNode) LoadFinalizedAndExecutedDispatcher(
+	node *NodeConfig,
+) (
+	module.ReadyDoneAware,
+	error,
+) {
+	latestFinalizedHeight :=
+		exeNode.builder.FinalizedHeader.Height
+
+	lastExecutedHeight, _, err :=
+		exeNode.executionState.GetHighestExecutedBlockID(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("cannot get the latest executed block height for "+
+			"the FinalizedAndExecutedDispatcher: %w", err)
+	}
+
+	latestExecutedAndFinalizedHeight := lastExecutedHeight
+	// take the minimum of the two heights
+	if latestFinalizedHeight < latestExecutedAndFinalizedHeight {
+		latestExecutedAndFinalizedHeight = latestFinalizedHeight
+	}
+
+	isBlockFinalized := func(ctx context.Context, h *flow.Header) (bool, error) {
+		// This uses the fact that only finalized headers can be retrieved by height.
+		finalizedHeader, err := node.Storage.Headers.ByHeight(h.Height)
+		if err != nil {
+			if errors.Is(err, storageerr.ErrNotFound) {
+				return false, nil
+			}
+			return false, err
+		}
+
+		return finalizedHeader.ID() == h.ID(), nil
```

Comment:
A much cheaper check is to compare the view. Because an executed block must have a QC, which means its view is unique as well. 
```suggestion
		return finalizedHeader.View == h.View, nil
```